### PR TITLE
MINOR: Ensure in-memory metadata is removed before physical deletion of segment

### PR DESCRIPTION
**Pull request by dhruvilshah3**
Minor refactoring of the places where we delete log segments, to ensure we always remove the in-memory metadata of the segment before performing physical deletion.

### CODE REVIEW

File: core/src/main/scala/kafka/log/Log.scala
Comment (rite2nikhil) [inline core/src/main/scala/kafka/log/Log.scala]: Instead of List[LogSegment] use Iterable[LogSegment]) ?

File: core/src/main/scala/kafka/log/Log.scala
Comment (rite2nikhil) [inline core/src/main/scala/kafka/log/Log.scala]: perhaps rename to deleteSegment ? also its singular ?

File: core/src/main/scala/kafka/log/Log.scala
Comment (rite2nikhil) [inline core/src/main/scala/kafka/log/Log.scala]: NIT: "from" is not required

**Review by rite2nikhil — CHANGES_REQUESTED** [review]

File: core/src/main/scala/kafka/log/Log.scala
Comment (dhruvilshah3) [reply core/src/main/scala/kafka/log/Log.scala]: I think using `List` is safer, to force callers to materialize the view of segments it derived from the `segments` map. We mutate the map when deleting the segment, which may cause the result of iterators derived from the map to be undefined.

File: core/src/main/scala/kafka/log/Log.scala
Comment (dhruvilshah3) [reply core/src/main/scala/kafka/log/Log.scala]: I wanted to make a distinction between `removeAndDeleteSegment` which performs a logical deletion and schedules a physical deletion vs this method which only performs physical deletion. Each segment consists of multiple files and thus `deleteSegmentFiles`. If this is confusing, I am open to ideas for a better name.

Comment (rite2nikhil) [reply core/src/main/scala/kafka/log/Log.scala]: makes sense, thanks

Comment (rite2nikhil) [reply core/src/main/scala/kafka/log/Log.scala]: I looked the code I see other related functions like mayDeleteSegment, deleteSegments, deleteOldSegments, it feels that renaming by adding "Files" throws the consistency away, calling it just deleteSegment makes sense to me

File: core/src/main/scala/kafka/log/Log.scala
Comment (rite2nikhil) [reply core/src/main/scala/kafka/log/Log.scala]: I am not sure I understand how using List makes it safer, one could really has a list or a array, it should not matter

File: core/src/main/scala/kafka/log/Log.scala
Comment (rite2nikhil) [inline core/src/main/scala/kafka/log/Log.scala]: nit : why extra line here (and in other places where this function is called ?

**Review by rite2nikhil — APPROVED** [review]
Thanks, changes overall look good,  see if any of my comments help improve.

File: core/src/main/scala/kafka/log/Log.scala
Comment (dhruvilshah3) [reply core/src/main/scala/kafka/log/Log.scala]: This is not a public API so I am not sure if the goal is to provide a generic API. Like I said, the intent is to materialize any iterators derived from the underlying collection, so we can safely mutate it to remove the segment. I could change to taking in an `Iterable` and then convert it to a list inside `removeAndDeleteSegments` if you think that is better.

Comment (rite2nikhil) [reply core/src/main/scala/kafka/log/Log.scala]: not needed, that would be cosmetic, my comment was on conversion of iterable to list when the underlying function is essentially iterating to delete the unordered members in the set

File: core/src/main/scala/kafka/log/Log.scala
Comment (dhruvilshah3) [reply core/src/main/scala/kafka/log/Log.scala]: I am not sure I follow. Are you suggesting the distinction between physical and logical deletion is not worth having? Note that a logical deletion eventually causes a physical deletion, so all the methods you pointed to would eventually call `deleteSegmentFiles`. I don't feel very strongly about the naming so I could change it back but I disagree that this throws the consistency away. As long as it is clear that we are invoking the method to perform both logical and physical deletion vs only physical deletion, I am okay with any naming.

File: core/src/main/scala/kafka/log/Log.scala
Comment (rite2nikhil) [inline core/src/main/scala/kafka/log/Log.scala]: where is asyncDelete == false used ?

File: core/src/main/scala/kafka/log/Log.scala
Comment (rite2nikhil) [reply core/src/main/scala/kafka/log/Log.scala]: i agree with you, it is better to have distinction, thanks for the explanation :)

File: core/src/main/scala/kafka/log/Log.scala
Comment (dhruvilshah3) [reply core/src/main/scala/kafka/log/Log.scala]: https://github.com/apache/kafka/pull/7106/files#diff-ffa8861e850121997a534ebdde2929c6R1965

**Comment by ijuma** [comment]
retest this please

**Comment by ijuma** [comment]
Thanks for the PR, the title should ideally describe the change. I'd probably change it to `Remove in-memory metadata of the segment before performing physical deletion` or something like that. It provides much more useful information than a generic message about refactoring.

File: core/src/main/scala/kafka/log/Log.scala
Comment (ijuma) [reply core/src/main/scala/kafka/log/Log.scala]: I'd suggest not having any default for `asyncDelete` to avoid unintentional bugs.

File: core/src/main/scala/kafka/log/Log.scala
Comment (ijuma) [inline core/src/main/scala/kafka/log/Log.scala]: Supposedly we don't need this anymore?

**Review by ijuma — COMMENTED** [review]
Left a few comments, it would be good for @junrao to take a deeper look.

File: core/src/main/scala/kafka/log/Log.scala
Comment (junrao) [inline core/src/main/scala/kafka/log/Log.scala]: Hmm, this doesn't seem right. We want to only delete from the first segment having truncatedBytes > 0, not necessarily all unflushedSegments.

File: core/src/main/scala/kafka/log/Log.scala
Comment (junrao) [inline core/src/main/scala/kafka/log/Log.scala]: Do we need this info logging given the logging in deleteSeg()?

**Review by junrao — COMMENTED** [review]
@dhruvilshah3 : Thanks for the PR. A couple of comments below.

File: core/src/main/scala/kafka/log/Log.scala
Comment (ijuma) [reply core/src/main/scala/kafka/log/Log.scala]: I think this one says when it's been scheduled versus when deletion started. Is that not useful?

Comment (junrao) [reply core/src/main/scala/kafka/log/Log.scala]: Hmm, it carries the same info just a bit early. Would that be useful?

File: core/src/main/scala/kafka/log/Log.scala
Comment (junrao) [reply core/src/main/scala/kafka/log/Log.scala]: Also, we need to short circuit the loop when we hit the first segment with truncatedBytes > 0.

File: core/src/main/scala/kafka/log/Log.scala
Comment (ijuma) [reply core/src/main/scala/kafka/log/Log.scala]: It could be useful if the scheduler gets overwhelmed and there's a long delay between scheduling and actual deletion. I don't feel strongly though, so leave it to you and @dhruvilshah3 to decide.

File: core/src/main/scala/kafka/log/Log.scala
Comment (dhruvilshah3) [reply core/src/main/scala/kafka/log/Log.scala]: Thanks for catching this! I caught this earlier today and pushed a fix in 509b916. I will push another commit to make the short circuit more explicit.

File: core/src/main/scala/kafka/log/Log.scala
Comment (dhruvilshah3) [reply core/src/main/scala/kafka/log/Log.scala]: > I'd suggest not having any default for asyncDelete to avoid unintentional bugs.

Sounds good. I made the change in 509b916.

File: core/src/main/scala/kafka/log/Log.scala
Comment (dhruvilshah3) [reply core/src/main/scala/kafka/log/Log.scala]: Yeah, removed.

File: core/src/main/scala/kafka/log/Log.scala
Comment (junrao) [inline core/src/main/scala/kafka/log/Log.scala]: To be consistent with removeAndDeleteSegments, perhaps segments should be Iterable?

File: core/src/main/scala/kafka/log/Log.scala
Comment (dhruvilshah3) [reply core/src/main/scala/kafka/log/Log.scala]: I don't feel strongly either. Most of the times, we care about when a segment was deleted. When the related files were deleted is perhaps not as interesting to know. In that case, we could log every time a segment is removed from the `segments` map?

File: core/src/main/scala/kafka/log/Log.scala
Comment (junrao) [inline core/src/main/scala/kafka/log/Log.scala]: Perhaps truncated is more meaningful than continue?

**Review by junrao — COMMENTED** [review]
@dhruvilshah3 : Thanks for the updated PR. Just a couple more minor comments.

File: core/src/main/scala/kafka/log/Log.scala
Comment (junrao) [inline core/src/main/scala/kafka/log/Log.scala]: It's probably better to log the segment deletion just before it's actually deleted. If we want to add the extra logging at the time of scheduling, it would be fine.

Comment (dhruvilshah3) [reply core/src/main/scala/kafka/log/Log.scala]: Hmm, I thought physical deletion of segments is not interesting and perhaps we could log logical deletions only. I will revert the logging to how it was before this PR.

**Review by junrao — APPROVED** [review]
@dhruvilshah3 : Thanks for the latest PR. LGTM. I will wail for the tests to complete before merging.

**Pull request merged**
